lib/snitch-core/src/lib.rs

```diff
@@ -0,0 +1,795 @@
+            return;
+        }
+
+        for _ in 0..=index {
+            // Discard the lead record in the queue.
+            let mut slices = self.storage.as_slices();
+            let len = u16::from_le_bytes(take_array(&mut slices).unwrap_lite())
+                as usize;
+
+            let size = OVERHEAD + len;
+            // So it's weird, but, heapless::Deque only lets you pop individual
+            // bytes. Hopefully this is not too expensive.
+            for _ in 0..size {
+                self.storage.pop_front();
+            }
+
+            self.stored_record_count -= 1;
+            self.earliest_ena += 1;
+        }
+    }
+
+    /// Makes a best effort at inserting a record.
+    ///
+    /// This will attempt to recover from the `Losing` state before writing
+    /// anything.
+    ///
+    /// If, after recovery, we are `Collecting` and there is enough space for a
+    /// header (`OVERHEAD`) plus `data`, the record is stored, implicitly
+    /// assigned the next ENA.
+    ///
+    /// If recovery is not possible, the existing loss count is advanced.
+    ///
+    /// If we are `Collecting` but `data` plus a header won't fit, we enter the
+    /// `Losing` state with a count of 1.
+    fn insert_impl(&mut self, sender: u16, timestamp: u64, data: &[u8]) {
+        let data_len = u16::try_from(data.len()).unwrap_lite();
+        self.recover_if_required(Some(OVERHEAD + data.len()));
+
+        match &mut self.insert_state {
+            InsertState::Collecting => {
+                let room = self.storage.capacity() - self.storage.len();
+                if room >= OVERHEAD + data.len() {
+                    self.write_header(sender, timestamp, data_len);
+                    for &byte in data {
+                        self.storage.push_back(byte).unwrap_lite();
+                    }
+                } else {
+                    self.insert_state = InsertState::Losing {
+                        count: NonZeroU32::new(1).unwrap_lite(),
+                        timestamp,
+                    };
+                }
+            }
+            InsertState::Losing { count, .. } => {
+                *count = count.saturating_add(1);
+            }
+        }
+    }
+
+    /// Internal utility routine for storing a record header.
+    fn write_header(&mut self, task: u16, timestamp: u64, data_len: u16) {
+        for byte in data_len.to_le_bytes() {
+            self.storage.push_back(byte).unwrap_lite();
+        }
+        for byte in task.to_le_bytes() {
+            self.storage.push_back(byte).unwrap_lite();
+        }
+        for byte in timestamp.to_le_bytes() {
+            self.storage.push_back(byte).unwrap_lite();
+        }
+        self.stored_record_count += 1;
+    }
+
+    /// Checks if we're losing data and attempts to stop, by generating a loss
+    /// record in the queue.
+    fn recover_if_required(&mut self, space_required: Option<usize>) -> bool {
```

Comment:
This method returns a `bool` but, as far as I can tell, no callers of it actually consume that `bool`. Is this vestigial, or are you expecting to need it later?

Reply:
I keep expecting to need it, and yet I haven't... I will flag it for reconsideration.